### PR TITLE
[Bugfix] Fix TieredSpilloverCache stats not adding correctly when shards are closed

### PR DESCRIPTION
**Pull request by peteralfonsi**
### Description
Fixes a bug where the total stats for the TieredSpilloverCache are decremented incorrectly when shards were closed. Misses and evictions from both the heap and disk tier were subtracted from the total, but this is incorrect. When the disk tier is enabled, only disk-tier misses and evictions should count towards the cache total, so only they should be subtracted. Adds UTs and ITs around this. 

Also adds UT coverage for TieredSpilloverCacheStatsHolder, which was missing before. 

### Related Issues
Resolves https://github.com/opensearch-project/OpenSearch/issues/16559

### Check List
- [x] Functionality includes testing.
- [N/A] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [N/A] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: modules/cache-common/src/main/java/org/opensearch/common/cache/stats/TieredSpilloverCacheStatsHolder.java [hunk 1]
```diff
@@ -6,9 +6,9 @@
  * compatible open source license.
  */
 
-package org.opensearch.cache.common.tier;
+package org.opensearch.common.cache.stats;
 
-import org.opensearch.common.cache.stats.DefaultCacheStatsHolder;
+import org.opensearch.common.cache.tier.TieredSpilloverCache;
```
Comment (sgup432) [inline modules/cache-common/src/main/java/org/opensearch/common/cache/stats/TieredSpilloverCacheStatsHolder.java:11]: Why do we need to include TSC class here?

File: modules/cache-common/src/main/java/org/opensearch/common/cache/stats/TieredSpilloverCacheStatsHolder.java [hunk 2]
```diff
@@ -164,7 +166,31 @@ public void decrementItems(List<String> dimensionValues) {
         super.decrementItems(dimensionValues);
     }
 
-    void setDiskCacheEnabled(boolean diskCacheEnabled) {
+    public void setDiskCacheEnabled(boolean diskCacheEnabled) {
         this.diskCacheEnabled = diskCacheEnabled;
     }
+
+    @Override
+    public void removeDimensions(List<String> dimensionValues) {
```
Comment (sgup432) [inline modules/cache-common/src/main/java/org/opensearch/common/cache/stats/TieredSpilloverCacheStatsHolder.java:174]: Can't we just reuse removeDimensions method from DefaultCacheStatsHolder? Both look exactly same.

File: CHANGELOG.md [hunk 3]
```diff
@@ -82,6 +82,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 - Fix warnings from SLF4J on startup when repository-s3 is installed ([#16194](https://github.com/opensearch-project/OpenSearch/pull/16194))
 - Fix protobuf-java leak through client library dependencies ([#16254](https://github.com/opensearch-project/OpenSearch/pull/16254))
 - Fix multi-search with template doesn't return status code ([#16265](https://github.com/opensearch-project/OpenSearch/pull/16265))
+- Fix bug in new cache stats API when closing shards while using TieredSpilloverCache ([#16560](https://github.com/opensearch-project/OpenSearch/pull/16560))
```
Comment (sgup432) [inline CHANGELOG.md:85]: Reword something to `[Tiered Caching] Fix bug in cache stats API`? Should be enough

File: modules/cache-common/src/internalClusterTest/java/org/opensearch/common/cache/tier/TieredSpilloverCacheBaseIT.java [hunk 4]
```diff
@@ -6,7 +6,7 @@
  * compatible open source license.
  */
 
-package org.opensearch.cache.common.tier;
+package org.opensearch.common.cache.tier;
```
Comment (sgup432) [inline modules/cache-common/src/internalClusterTest/java/org/opensearch/common/cache/tier/TieredSpilloverCacheBaseIT.java:9]: Can we avoid renaming these packages? This has introduced a lot of changes.
Plus for `cache-common` module, it makes sense to keep package name prefix as `org.opensearch.cache`

File: modules/cache-common/src/main/java/org/opensearch/common/cache/stats/TieredSpilloverCacheStatsHolder.java (hunk 2 above)
Comment (peteralfonsi) [reply modules/cache-common/src/main/java/org/opensearch/common/cache/stats/TieredSpilloverCacheStatsHolder.java:174]: For the TSC we check for `assert dimensionValues.size() == dimensionNames.size() - 1` rather than `dimensionNames.size()`, so that the base case is the node whose children are the individual tier nodes, but besides the check they're otherwise the same.

File: modules/cache-common/src/internalClusterTest/java/org/opensearch/common/cache/tier/TieredSpilloverCacheBaseIT.java (hunk 4 above)
Comment (peteralfonsi) [reply modules/cache-common/src/internalClusterTest/java/org/opensearch/common/cache/tier/TieredSpilloverCacheBaseIT.java:9]: We can do this if we make a couple methods in `DefaultCacheStatsHolder` public. Otherwise I'm not sure the tests can work out. Will make these changes.

File: modules/cache-common/src/main/java/org/opensearch/common/cache/stats/TieredSpilloverCacheStatsHolder.java (hunk 1 above)
Comment (peteralfonsi) [reply modules/cache-common/src/main/java/org/opensearch/common/cache/stats/TieredSpilloverCacheStatsHolder.java:11]: Think something got automatically imported when I had refactored the packages since they ended up in different packages; now that I've reverted this change this import isn't needed anymore.

File: modules/cache-common/src/main/java/org/opensearch/cache/common/tier/TieredSpilloverCache.java [hunk 5]
```diff
@@ -373,14 +373,14 @@ private V compute(ICacheKey<K> key, LoadAwareCacheLoader<ICacheKey<K>, V> loader
...
-                    statsHolder.removeDimensions(dimensionValues);
-                }
-                if (key.key != null) {
-                    try (ReleasableLock ignore = writeLock.acquire()) {
-                        cacheEntry.getKey().invalidate(key);
+            if (key.getDropStatsForDimensions()) {
+                statsHolder.removeDimensions(key.dimensions);
+            } else {
+                for (Map.Entry<ICache<K, V>, TierInfo> cacheEntry : caches.entrySet()) {
+                    if (key.key != null) {
```
Comment (andrross) [inline modules/cache-common/src/main/java/org/opensearch/cache/common/tier/TieredSpilloverCache.java:380]: Can the null check be moved outside the for loop?

File: server/src/main/java/org/opensearch/common/cache/stats/DefaultCacheStatsHolder.java [hunk 6]
```diff
@@ -209,14 +211,14 @@ private ImmutableCacheStats removeDimensionsHelper(List<String> dimensionValues,
     }
 
     // pkg-private for testing
```
Comment (andrross) [inline server/src/main/java/org/opensearch/common/cache/stats/DefaultCacheStatsHolder.java:213]: Comment is no longer accurate

File: modules/cache-common/src/main/java/org/opensearch/cache/common/tier/TieredSpilloverCacheStatsHolder.java [hunk 7]
```diff
@@ -167,4 +169,17 @@ public void decrementItems(List<String> dimensionValues) {
     void setDiskCacheEnabled(boolean diskCacheEnabled) {
         this.diskCacheEnabled = diskCacheEnabled;
     }
+
+    @Override
+    public void removeDimensions(List<String> dimensionValues) {
```
Comment (andrross) [inline modules/cache-common/src/main/java/org/opensearch/cache/common/tier/TieredSpilloverCacheStatsHolder.java:174]: This method looks identical to the method it is overriding in the parent class. Why was this added?

Comment (peteralfonsi) [reply modules/cache-common/src/main/java/org/opensearch/cache/common/tier/TieredSpilloverCacheStatsHolder.java:174]: The only difference is the assertion checking the list of values passed in is `dimensionNames.size() - 1` instead of `dimensionNames.size()`, but the logic is otherwise the same. We could remove this assertion, and replace the overridden method with a comment somewhere, but my thinking was that the explicit check makes it clearer to developers how to use `removeDimensions` for the inheriting class.

Comment (peteralfonsi) [reply modules/cache-common/src/main/java/org/opensearch/cache/common/tier/TieredSpilloverCacheStatsHolder.java:174]: Happy to remove it if you think a comment is sufficient

Comment (peteralfonsi) [reply modules/cache-common/src/main/java/org/opensearch/cache/common/tier/TieredSpilloverCacheStatsHolder.java:174]: @andrross, please let me know what you think on this, happy to do either way.

**Review by andrross — APPROVED** [review]
